### PR TITLE
fetch ws state from IPFS via cli

### PR DESCRIPTION
**Pull request by 3xtr4t3rr3str14l**
Changes our cli to have it load from a weak subjectivity state by default (if a state is not already in the db)
Also adds a state root file of the corresponding ws state to verify against
As it stands right now with this scheme, we will have to update the IPFS link and the state root file manually for every release.

The current ws state is stored here:
https://ipfs.io/ipfs/QmVsGxwZs4acTDfhQrNYVqccEVxyhALqJuYqgJrbbNAd1Z

This is all for the Prater testnet.  I can generate one for mainnet as well once this PR and the parallel PR for beacon-state-upload (https://github.com/ChainSafe/beacon-state-upload/pull/3) are merged to their respective master branches.

**NOTE: DO NOT MERGE:** added the DO NOT MERGE label because we need both this PR and https://github.com/ChainSafe/beacon-state-upload/pull/3 to be approved before merging, but also because the downloading of the state from IPFS isn't always stable during my testing, so I want to investigate this further.

### CODE REVIEW

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (wemeetagain) [inline packages/cli/src/cmds/beacon/initBeaconState.ts]: Can we pull out the ipfs path and state root into a separate file, eg: weakSubjectivityState.json.
Something that ends up looking like:
```json
{
  "ipfsPath": "/ipfs/QmVsGxwZs4acTDfhQrNYVqccEVxyhALqJuYqgJrbbNAd1Z/state.ssz",
  "stateRoot": "0xabcd..."
}
```

I think this will be easier to manage.

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (wemeetagain) [inline packages/cli/src/cmds/beacon/initBeaconState.ts]: maybe missing "IPFS" in this error message?

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (dapplion) [inline packages/cli/src/cmds/beacon/initBeaconState.ts]: Why would this file `./weakSubjectivityState.json` exist?

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (dapplion) [inline packages/cli/src/cmds/beacon/initBeaconState.ts]: The IPFS gateway url must be a customizable parameter

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (3xtr4t3rr3str14l) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: as it stands right now, it would be because we copied it from the output of beacon-state-upload (from here: https://github.com/ChainSafe/beacon-state-upload/pull/3/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R39)
perhaps we can have a more automated process, but that's the way it is right now (or at least inasmuch as this PR and that beacon-state-upload PR are "right now")

**Comment by 3xtr4t3rr3str14l** [comment]
added the DO NOT MERGE label because we need both this PR and https://github.com/ChainSafe/beacon-state-upload/pull/3 to be approved before merging, but also because the downloading of the state from IPFS isn't always stable during my testing, so I want to investigate this further.

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (dapplion) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: I still don't follow why the file should exist in any environment that you may run the cli.

Comment (3xtr4t3rr3str14l) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: Two reasons
1) so we can have the state root of the weak subjectivity state to check against when downloading from IPFS so as to not rely purely on the security of the SSL certs (as discussed on one of the eth2 implementer's calls about Teku's state loading process from Infura)
2) so we can have the IPFS link from which to download the weak subjectivity state

Maybe it's better to just embed directly into the code though rather than rely on a file.  We'll just have to remember to update the code for every release that we have a new weak subjectivity state, but that's just as much of a process as remembering to copy the file anyway.

Comment (dapplion) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: But again if we add the file at the project root `/weakSubjectivityState.json`, and you run the command from the root it will work. If you run the command from `/packages/cli` it will break since `./weakSubjectivityState.json` does not exist

Comment (3xtr4t3rr3str14l) [reply packages/cli/src/cmds/beacon/initBeaconState.ts]: ahh ok i gotcha.  yeah i'll move it into `packages/lodestar/src/cmds/...`

**Comment by 3xtr4t3rr3str14l** [comment]
> added the DO NOT MERGE label because we need both this PR and [ChainSafe/beacon-state-upload#3](https://github.com/ChainSafe/beacon-state-upload/pull/3) to be approved before merging, but also because the downloading of the state from IPFS isn't always stable during my testing, so I want to investigate this further.

To follow up to this comment, I have been running the beacon state uploader for the past few days with no issues on mainnet.  Currently setting one up on contabo-3 to run alongside our prater node as well, but I expect similar results.  Thus, I feel that this PR is working well enough to where I don't need to have the DO NOT MERGE label anymore.

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (dapplion) [inline packages/cli/src/cmds/beacon/initBeaconState.ts:41]: You could verify it but it is invalid. You should print the expect root and the actual root

File: packages/cli/src/cmds/beacon/initBeaconState.ts
Comment (dapplion) [inline packages/cli/src/cmds/beacon/initBeaconState.ts:82]: Can be simplified by getting the `weakSubjectivityState` and then calling 
```ts
return await initAndVerifyWeakSujectivityState(args, weakSubjectivityState, initFromFile, logger);
```
once

File: packages/cli/src/cmds/weakSubjectivityState.ts
Comment (dapplion) [inline packages/cli/src/cmds/weakSubjectivityState.ts:16]: How would we update this values every X time?

Comment (dapplion) [reply packages/cli/src/cmds/weakSubjectivityState.ts:16]: Who is pinning those IPFS paths and with what replication factor?

**Comment by 3xtr4t3rr3str14l** [comment]
@dapplion just pulled a draft PR for a new version of the weak subjectivity CLI here:
https://github.com/ChainSafe/lodestar/pull/2571

Keeping this one (#2451) open for now so we can still have both solutions available to review, if needed.

File: packages/cli/src/cmds/weakSubjectivityState.ts
Comment (dapplion) [reply packages/cli/src/cmds/weakSubjectivityState.ts:16]: @3xtr4t3rr3str14l bump. This is the most critical question to answer around this PR

Comment (3xtr4t3rr3str14l) [reply packages/cli/src/cmds/weakSubjectivityState.ts:16]: Under this design, the pinning would be done by https://github.com/ChainSafe/beacon-state-upload/pull/3 every time a new weak subjectivity period is fetched via our API.  `beacon-state-upload` would be subscribed to the finalized checkpoints events stream via the API, and for every new finalized checkpoint, we would check to see if the WS API (via `lodestar/ws_epoch` custom API endpoint) gives us a new WS checkpoint.  If we have a new WS checkpoint, then we get the WS state from the API, and upload the WS checkpoint data and the WS state to IPFS & IPNS (which pins the data automatically via the IPFS HTTP API).  But under this setup, we would only use the IPFS link in this embedded file, for stability.

However, none of this may be relevant if we go with the newer solution in #2571 .  I'm also open to suggestions on a fusion of the two strategies though, if that is desired.

**Comment by wemeetagain** [comment]
closing in favor of #2571 which is nearing completion

**Pull request closed without merging**
